dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp

```diff
@@ -676,14 +696,25 @@ void DAGQueryBlockInterpreter::executeImpl(DAGPipeline & pipeline)
         recordProfileStreams(pipeline, query_block.limit_or_topn_name);
     }
 
-    // execute final project action
-    executeProject(pipeline, final_project, "final projection");
     // execute limit
     if (query_block.limit_or_topn && query_block.limit_or_topn->tp() == tipb::TypeLimit)
     {
         executeLimit(pipeline);
         recordProfileStreams(pipeline, query_block.limit_or_topn_name);
     }
+
+    // execute the expand OP after all filter/limits and so on.
+    // since expand OP has some row replication work to do, place it after limit can reduce some unnecessary burden.
+    // and put it before the final projection, because we should recognize some base col as grouping set col before change their alias.
+    if (res.before_expand)
+    {
+        executeExpand(pipeline, res.before_expand);
+        recordProfileStreams(pipeline, query_block.expand_name);
+    }
+
+    // execute final project action
+    executeProject(pipeline, final_project, "final projection");
```

Comment:
In the original implementation, the order was `final projection --> limit`.
How about maintaining this order? `expand --> final projection --> limit`

Reply:
expand OP is resource costly, i wanna it to be after limit to avoid redundant data expansion.

Reply:
```
std::pair<bool, BoolVec> DAGExpressionAnalyzer::isCastRequiredForRootFinalProjection(
    const std::vector<tipb::FieldType> & require_schema,
    const std::vector<Int32> & output_offsets) const
{
    /// TiDB can not guarantee that the field type in DAG request is accurate, so in order to make things work,
    /// TiFlash will append extra type cast if needed.
    const auto & current_columns = getCurrentInputColumns();
    bool need_append_type_cast = false;
    BoolVec need_append_type_cast_vec;
    /// we need to append type cast for root final projection if necessary
    for (UInt32 i : output_offsets)
    {
        const auto & actual_type = current_columns[i].type;

```
when build expand it will change the source-column(appending one more column), and if expand building is after final-proj, the offset is >= source column

Reply:
I'm not saying let expand come after the final projection, but rather, let limit come after the final projection and stay as it is

Reply:
And then `gtest_interpreter.out` will not be changed.

Reply:
got, my concern is: expand OP has some row replication work to do, place it after limit can reduce some unnecessary replication burden.

Reply:
my new commit has split before_order_and_select  into two parts, and place expand among them, otherwise, some old test fails

here is the new sequence: `before_order <- limit/topN <- expand <- before_select <- final-proj`

Reply:
ok, lgtm.